cuda/test/multigrid/amgx_pgm_kernels.cpp

```diff
@@ -98,13 +98,35 @@ class AmgxPgm : public ::testing::Test {
             ref);
     }
 
+    gko::Array<index_type> gen_agg_array(gko::size_type num,
+                                         gko::size_type num_agg)
+    {
+        auto agg_array = gen_array(num, 0, num_agg - 1);
+        auto agg_array_val = agg_array.get_data();
+        std::vector<index_type> select_agg(num);
+        std::iota(select_agg.begin(), select_agg.end(), 0);
+        // use the first num_agg item as the aggregated index.
+        std::shuffle(select_agg.begin(), select_agg.end(), rand_engine);
+        // the value of agg_array is the i-th of aggregate group
+        for (gko::size_type i = 0; i < num; i++) {
+            agg_array_val[i] = select_agg[agg_array_val[i]];
```

Comment:
Couldn't it happen that this is also equal to `i`, thus breaking your assumption for `num_agg == n` below?

Reply:
I think it should be fine.
the array contains at most num_agg distinct value.
later operation set the aggregated index such that the array contains exactly num_agg distinct value.
each group may has 1 ~ x elements

Reply:
if select_agg has a fixed point `select_agg[i] == i` larger than num_agg, wouldn't this lead to `num_agg + 1` distinct values?

Reply:
if i > num_agg, select_agg[i] will not be used. agg_array has 0 ~ num-1 value, so it will use the first num item from select_agg

Reply:
Oh yes sorry, I didn't realize that agg_array only contains values from 0 to num_agg - 1.